### DISCUSSION TITLE
Does it support wide angle camera ?

### DISCUSSION THREAD

**Discussion by florian-mlr** [pedroSG94/RootEncoder#1458] [Q&A]
Hello Pedro,

Does your lib support wide angle camera ?

From my research, detecting the wide angle on Android is not simple.
There are no "wide angle" capabilities strictly speaking.
Some devices have a physical wide angle camera, sometimes it can be listed in the cameras according to the focal length, depending on the manufacturers.
Devices supporting wide angle logically are based on the zoom level

In your opinion, what would be the tasks to implement a feature that supports wide angle with your library ?

**Comment by pedroSG94** [comment]
Hello,

Currently you can open cameras using the camera id and list cameras so if you can detect a physical wide angle camera with camera2 api you can do it:
https://github.com/pedroSG94/RootEncoder/blob/master/library/src/main/java/com/pedro/library/base/Camera2Base.java#L849
https://github.com/pedroSG94/RootEncoder/blob/master/library/src/main/java/com/pedro/library/base/Camera2Base.java#L514

I haven't a way to detect if the device has wide angle camera but theoretically you can check the focal of a camera so maybe we can find a wide angle camera using this way. We should get a device with a physical wide angle camera and get all focal length of the cameras.
We will need modify this class adding the logic for it:
https://github.com/pedroSG94/RootEncoder/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java
And then expose the method in Camera2Base and Camera2Source classes

But I have a question:
Which is the minimal focal length for a wide angle camera?

**Reply by florian-mlr** [reply]
Thank you for your answer.
About your question, i don't have reliable source but I read that a wide focal length would be between 12mm and 16mm

**Reply by florian-mlr** [reply]
I've create a function to get camera id for wide angle , the formula for wide angle comes from : https://antongorlin.com/blog/effective-focal-length-calculator/
 
```
fun getCameraIdForWideAngle(cameraManager: CameraManager): String? {
            for (cameraId in cameraManager.cameraIdList) {
                try {
                    val focalLengths = cameraManager.getCameraCharacteristics(cameraId)
                        .get(CameraCharacteristics.LENS_INFO_AVAILABLE_FOCAL_LENGTHS)?.firstOrNull()
                        ?: throw Exception("Can't get focal length for camera : $cameraId")

                    val sensor = cameraManager.getCameraCharacteristics(cameraId)
                        .get(CameraCharacteristics.SENSOR_INFO_PHYSICAL_SIZE)
                        ?: throw Exception("Can't get sensor physical size for camera : $cameraId")

                    val size = sqrt(sensor.width * sensor.width + sensor.height * sensor.height)

                    val effectiveFocalLength = focalLengths * (43.27 / size)

                    Timber.d("cameraId : $cameraId effectiveFocalLength : ${effectiveFocalLength}")

                    if (effectiveFocalLength in 12f..16f) {
                        return cameraId
                    }
                } catch (e: CameraAccessException) {
                    Timber.e(e)
                } catch (e: Exception) {
                    Timber.e(e)
                }
            }
            return null
        }
```

**Comment by pedroSG94** [comment]
Hello,

Thank you for the method. I added the implementation here:
https://github.com/pedroSG94/RootEncoder/pull/1459
Can you test it with your device and tell me if all is working fine?

**Reply by florian-mlr** [reply]
Hi Pedro, I will do a test next week and I will tell you.
Thanks for this quick implementation.
I am not yet sure of the 12 and 16mm values for the wide angle but I will continue my research on it.

**Reply by pedroSG94** [reply]
Ok,
I tested it but I have no cameras with that values.
I have values between 23-25
